api/routes/videos.go

```diff
@@ -1,72 +1,144 @@
-		if err := result.Error; err != nil {
-			w.WriteHeader(http.StatusNotFound)
-			w.Write([]byte("404"))
-			return
-		}
+func handleVideoRequest(
+	w http.ResponseWriter,
+	r *http.Request,
+	db *gorm.DB,
+	mediaName string,
+	authenticateFn func(*models.Media, *gorm.DB, *http.Request) (bool, string, int, error),
+	getCachePathFn func(albumID, mediaID int, filename string) string,
+) {
+	var mediaURLs []models.MediaURL
+	if err := db.Model(&models.MediaURL{}).
+		Preload("Media").
+		Where("media_urls.media_name = ? AND media_urls.purpose = ?", mediaName, models.VideoWeb).
+		Find(&mediaURLs).
+		Error; err != nil || len(mediaURLs) == 0 || mediaURLs[0].Media == nil {
 
-		var media = mediaURL.Media
+		w.WriteHeader(http.StatusNotFound)
+		w.Write([]byte("not found"))
+		return
+	}
 
-		if success, response, status, err := authenticateMedia(media, db, r); !success {
-			if err != nil {
-				log.Printf("WARN: error authenticating video: %s\n", err)
+	if len(mediaURLs) > 1 {
+		sanitizedMediaName := strings.Map(func(r rune) rune {
+			if unicode.IsPrint(r) {
+				return r
 			}
-			w.WriteHeader(status)
-			w.Write([]byte(response))
-			return
+			return -1
+		}, mediaName)
+		log.Warn(r.Context(), "Multiple video web URLs found",
+			"name", sanitizedMediaName,
+			"count", len(mediaURLs),
+			"using", mediaURLs[0],
+		)
+	}
+
+	mediaURL := mediaURLs[0]
+	var media = mediaURL.Media
+
+	if success, response, status, err := authenticateFn(media, db, r); !success {
+		if err != nil {
+			log.Warn(r.Context(), "got error authenticating video:",
+				"error", err,
+				"media ID", media.ID,
+				"media path", media.Path)
 		}
+		w.WriteHeader(status)
+		w.Write([]byte(response))
+		return
+	}
 
-		var cachedPath string
+	var cachedPath string
+
+	if mediaURL.Purpose == models.VideoWeb {
+		// Use the provided cache path function
+		cachedPath = getCachePathFn(int(media.AlbumID), int(mediaURL.MediaID), mediaURL.MediaName)
+	} else {
+		log.Error(r.Context(), "Can not handle media_purpose for video",
+			"purpose", mediaURL.Purpose,
+			"expected", models.VideoWeb)
+		w.WriteHeader(http.StatusInternalServerError)
+		w.Write([]byte(internalServerError))
+		return
+	}
+
+	if _, err := os.Stat(cachedPath); err != nil {
+		if os.IsNotExist(err) {
+			if err := processSingleMediaFn(r.Context(), db, media); err != nil {
+				// Check if error was due to context cancellation
+				if r.Context().Err() != nil {
+					log.Warn(r.Context(), "video processing cancelled due to client disconnect",
+						"mediaID", media.ID,
+						"reason", r.Context().Err())
+					return // Don't send response if client disconnected
+				}
 
-		if mediaURL.Purpose == models.VideoWeb {
-			cachedPath = path.Join(utils.MediaCachePath(), strconv.Itoa(int(media.AlbumID)), strconv.Itoa(int(mediaURL.MediaID)), mediaURL.MediaName)
+				log.Error(r.Context(), "processing video not found in cache:",
```

Comment:
In a structed log, it's not a one line. change the message to `processing video not found in cache`, no `:` in the suffix. See [slog details](https://pkg.go.dev/log/slog#pkg-overview)